Full text search LINQ support

rwasef1830 opened the pull request:
Hello,
This resolves #1 and implements full text LINQ query support with all methods and operators. It also adds an explicit casting function (EF.Functions.Cast) for edge cases (when needed). This brings it in feature parity with the EF6 provider.

Thanks @roji for the hints along the way. Hoping for a review and merge 👍

rwasef1830 commented:
Regressions found. Investigating...

rwasef1830 commented:
Was forced to drop array_to_tsvector as there are some internal assumptions in EF Core's expression processing that depend on array expressions not getting server evaluated and it is too fragile to try to distinguish between cases where the expressions can be supported server-side or not. So I decided to drop its support. array_to_tsvector is seldom used anyway, it is a utility function to join lexemes into a single expression.

Ready for review @roji

rwasef1830 commented:
Rebased on latest dev and squashed full text support with dropping array_to_tsvector

roji commented:
Great, so should I review?

rwasef1830 commented:
Yup!

roji commented:
OK, will make this the next thing on my plate after I finish my current task, promise.

roji commented on src/EFCore.PG/Query/Expressions/Internal/FullTextSearchExpression.cs:
```diff
@@ -0,0 +1,118 @@
+            ? $"{Left} {Operator} {Right}"
+            : $"{Operator} {Left}";
+
+        public static FullTextSearchExpression TsQueryAnd([NotNull] Expression left, [NotNull] Expression right)
```
I'm not sure this actually have much value... If/when you you refactor the expression as I suggested, creating a "contains" expression would be a matter of instantiating a FullTextSearchBinaryExpression and passing an enum to the constructor (`NpgsqlFullTextOperation.Contains` or something similar).

roji commented on src/EFCore.PG/Storage/Internal/Mapping/NpgsqlTsQueryTypeMapping.cs:
```diff
@@ -0,0 +1,24 @@
+            query.Write(builder, true);
+            builder.Replace("'", "''");
+            builder[0] = '\'';
+            builder.Append("'::tsquery");
```
The literal representations as usually of the form `HSTORE 'blablabla'` rather than `'blablabla'::hstore`, it's probably a good idea to align to that.

rwasef1830 replied:
Hmmm I didn't know this syntax. I will update it to match.

roji commented on src/EFCore.PG/Storage/Internal/Mapping/NpgsqlTsVectorMapping.cs:
```diff
@@ -0,0 +1,24 @@
+            builder.Append(vector);
+            builder.Replace("'", "''");
+            builder[0] = '\'';
+            builder.Append("'::tsvector");
```
Same as above

rwasef1830 replied:
Will change it.

rwasef1830 commented on src/EFCore.PG/NpgsqlDbFunctionsExtensions.cs:
```diff
@@ -159,5 +159,11 @@ var regexPattern
+        /// Casts the specified <paramref name="target" /> to the PostgreSQL type equivalent
+        /// to <typeparamref name="TTarget" />.
+        /// </summary>
+        public static TTarget Cast<TTarget>(this DbFunctions _, object target) => throw new NotSupportedException();
```
It's a useful utility for emitting a cast on the PostgreSQL side that may not have a conversion operator or method defined on the C# side and thus may not be expressible in LINQ directly.

I have a similar method in the EF6 provider, so this is for feature parity. It is not strictly needed for full text search but may be useful for others.

roji replied:
I can see the usefulness of this, but it seems like something that belongs in EF Core rather than Npgsql - it has really nothing to do with this specific provider. EFCore.Relational has some "universal" relational translators such as for Like (as well as some others), and it would make sense to have a Cast translator in there as well.

One more note... in this method you're providing the cast target as a CLR type (`TTarget`), which is great. But in some cases I found myself needing to specify a store datatype as a string: when you specify as a CLR type, you're going through the provider's type mappings, which in effect limits you to the default store datatype for that CLR type. For example, there's no way to translate to PostgreSQL type `jsonb`, since its CLR type is string, but the default datatype for that is actually `text`.

I recently had this issue when fixing #319, where I needed to cast to `citext`, and added my own `ExplicitStoreTypeCastExpression`, which is the same as an `ExplicitCastExpression` but which is provided a store datatype as a string instead of a CLR type, effectively bypassing the type mappings. This also probably belongs in EF Core rather than the provider.

If you end up doing a PR to EF Core, you may want to add the string-based cast option as well.

ajcvickers replied:
@roji Seems related to https://github.com/aspnet/EntityFrameworkCore/issues/4978 which started off being about Unicode, but in general is about providing type mappings for parts of a query where it is not possible/easy to infer the mapping from some related property. /cc @divega

roji replied:
Thanks @ajcvickers. I do think it also makes sense to provide some means for users to force a cast sometimes, as a last resort or workaround. As I said above, I ended up needing a manual explicit cast to a store type (not a CLR type - to bypass the type mapping system), you can [check this code out](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/5488caf054f3b56c2b8ab33b34221d6aeffa8619/src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlStringStartsWithTranslator.cs#L73).

divega replied:
@roji @rwasef1830 @ajcvickers still don't understand how `EF.Functions.Cast<X>(y)` would be different form `(X)y`. Is it different because the latter may be removed or ignored by EF Core?

rwasef1830 replied:
```(X)y``` will not compile if there is no C# implementation of a cast operator or a built-in conversion. The method call will be allowed to compile. There is no difference in case there *is* a conversion, but that's not the point of the method.

ajcvickers replied:
@divega See https://github.com/aspnet/EntityFrameworkCore/issues/11089 where the answer was to force an "invalid" cast into the model using `(string)(object)myInt`. This works because it is never executed, but the cast call seems cleaner. But my main point was that something like this could allow facets to be changed.

roji replied:
Keep in mind that there are also store types which simply don't exist in C#. For example, CLR string maps to both text and json in PostgreSQL, with text being the default. But in some scenarios I may need to specify a cast to JSON specifically (since PostgreSQL doesn't do implicit casting in several scenarios).

divega replied:
Ok, thanks, I think get it. `EF.Functions.Cast<X>(y)` is useful if the three conditions below are true:
1. There is no valid conversion
2. You don't want to write `(X)(object)y`
3. The X CLR type provides enough information (e.g. it is not about explicitly casting to store type or specifying a facet

I think it is interesting, and we could have something like this in EF Core, but perhaps it is more interesting to consider other overloads or variations that can set various facets or the store type. I think it is unlikely that adding this now would prevent us from adding the other variations we want later.

roji replied:
@rwasef1830, I guess you want to open an issue for this feature in the EF Core repo, and remove this commit from the PR?

rwasef1830 replied:
OK, got caught up and busy at the moment, will get it done next week hopefully.

rwasef1830 commented on src/EFCore.PG/Query/Internal/NpgsqlEvaluatableExpressionFilter.cs:
```diff
@@ -11,11 +13,23 @@ public class NpgsqlEvaluatableExpressionFilter : RelationalEvaluatableExpression
         public override bool IsEvaluatableMethodCall(MethodCallExpression methodCallExpression) =>
             (!methodCallExpression.Method.IsGenericMethod
              || methodCallExpression.Method.GetGenericMethodDefinition() != _cast)
+            && methodCallExpression.Method != _tsQueryParse
```
This is because EF Core tries to compile the LINQ query and does a first pass optimization using ParameterExtractingExpressionVisitor eliminating expressions that do not reference any columns and evaluating them on the client side and sending the results directly as a literal.

Without this my unit tests break and it introduces a subtle behavior anomaly where if you do something like this:

```
context.Query<SomeEntity>(x => 1 + 1);
```
The 1+1 expression will be evaluated client-side and not translated to SQL.

But if you do this:
```
context.Query<SomeEntity>(x => x + 1);
```
It will be properly be translated to SQL.

Entity Framework Core people were aware of this issue from before and had used a workaround for this (FunctionEvaluationDisablingVisitor) to avoid re-linq doing this optimization when not desired, but later on a native way was introduced in re-linq (EvaluatableExpressionFilter) so they removed that visitor and used the new way instead.

Note this affects your Regex implementation as well, you may want to add Regex methods to the filter.

roji replied:
Thanks for explaining this. The thing still causing me trouble, is why this isn't happening in EF Core, the SQL Server provider or anywhere else I can see.

EF Core's [base EvaluatableExpressionFilter](https://github.com/aspnet/EntityFrameworkCore/blob/dev/src/EFCore/Query/Internal/EvaluatableExpressionFilter.cs) only lists a handful of non-deterministic methods such as `DateTime.Now`, while [RelationalEvaluatableExpressionFilter](https://github.com/aspnet/EntityFrameworkCore/blob/dev/src/EFCore.Relational/Query/Internal/RelationalEvaluatableExpressionFilter.cs) adds user-defined database functions to that. The SQL Server and Sqlite providers don't have a EvaluatableExpressionFilter at all.

In other words, the entire rich array of methods that EF Core knows how to translate (including SQL Server's `FREETEXT` method (full-text search) is absent from EvaluatableExpressionFilter. Any idea why that is? We can also ask the EF Core team about this.

Regarding the Regex implementation, isn't it actually an advantage that an expression gets evaluated locally when possible? I mean, if the user writes an expression with regex that *doesn't* reference any database columns (this would be weird), why should that get sent to the database for evaluation?

divega replied:
@smitpatel, @anpete there are some good query design questions here that I don't feel prepared to answer accurately... Can you weigh in? 

In general I do agree with @roji that we prefer evaluating things that are uncorrelated to the server on the client.

Also, not sure you care but this is probably not a safe way to compare reflection objects across all .NET implementations.

anpete replied:
> why this isn't happening in EF Core, the SQL Server provider or anywhere else I can see.

Because, I think, these new functions are unusual in that they may be used in a way where they are not correlated to any range variable in the query (at least that's what it looks like from looking at the tests). In the more normal case, the function call expression contains one or more range variable expressions, and so would never be considered for partial eval.

We may want to allow providers to simply add entries to the base EvaluatableExpressionFilter in order to avoid a service replacement here..

roji replied:
@anpete I don't think the these new functions are very special - it's true that the tests currently seem to exercise mostly local data, but the functions obviously make sense mostly with database data (after all we're talking about full-text search)...

Regardless, definitely a +1 for allowing adding entries in EvaluatableExpresisonFilter (like the member/method translators?) - and so also moving it out of Internal.

rwasef1830 commented on src/EFCore.PG/FullTextSearch/NpgsqlFullTextSearchDbFunctionsExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+using System.Diagnostics.CodeAnalysis;
+using NpgsqlTypes;
+
+namespace Microsoft.EntityFrameworkCore.FullTextSearch
```
Ok I will change this.

rwasef1830 commented on src/EFCore.PG/FullTextSearch/NpgsqlFullTextSearchLinqExtensions.cs:
```diff
@@ -0,0 +1,269 @@
+using System.Diagnostics.CodeAnalysis;
+using NpgsqlTypes;
+
+namespace Microsoft.EntityFrameworkCore.FullTextSearch
```
Ok.

rwasef1830 commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlFullTextSearchMethodTranslator.cs:
```diff
@@ -0,0 +1,287 @@
+{
+    public class NpgsqlFullTextSearchMethodTranslator : IMethodCallTranslator
+    {
+        private static readonly MethodInfo _tsQueryParse = typeof(NpgsqlTsQuery).GetMethod(
```
Some parts of the codebase omit it and others keep it, I saw in the parts that I changed that the private modifier was being used, so I used it explicitly to keep in line.

If you prefer to remove it, I will remove it from the code in this pull request, but the entire codebase needs to be combed for this issue.

rwasef1830 commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlFullTextSearchMethodTranslator.cs:
```diff
@@ -0,0 +1,287 @@
+            }
+
+            if (methodCallExpression.Method.DeclaringType == typeof(NpgsqlFullTextSearchDbFunctionsExtensions))
+            {
```
I personally like to only omit curlies when the expression is short enough to be on the same line as the if condition. Once the condition moves to its own line, I like to use the curlies to make it clearer. Anyway, I will change it.

rwasef1830 commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlFullTextSearchMethodTranslator.cs:
```diff
@@ -0,0 +1,287 @@
+        {
+            switch (methodCallExpression.Method.Name)
+            {
+            case nameof(NpgsqlFullTextSearchLinqExtensions.And):
```
To avoid littering the class with tons of MethodInfo fields for the many overloads.

rwasef1830 commented on src/EFCore.PG/Query/Expressions/Internal/FullTextSearchExpression.cs:
```diff
@@ -0,0 +1,118 @@
+
+namespace Microsoft.EntityFrameworkCore.Query.Expressions.Internal
+{
+    public class FullTextSearchExpression : Expression
```
Doh! You're right I totally missed that. I guess I was thinking of keeping full text stuff separate and explicit to keep it readable, I will refactor it.

rwasef1830 replied:
Actually for ```@@```, ```@>``` and ```@<``` I'm still forced to create a top-level expression because ```BinaryExpression``` can't be inherited from (its constructors are internal).

roji replied:
You're right, I missed that. It's a shame, that would have led to a very clean design.

rwasef1830 commented on src/EFCore.PG/FullTextSearch/NpgsqlFullTextSearchDbFunctionsExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+        /// by <paramref name="config" />.
+        /// http://www.postgresql.org/docs/current/static/textsearch-controls.html#TEXTSEARCH-PARSING-DOCUMENTS
+        /// </summary>
+        public static NpgsqlTsVector ToTsVector(this DbFunctions _, string config, string document) =>
```
I thought about this, but since both parameter types are strings, I thought it will trip up people who know the full text functions and cause confusion and hate :-D, so I preferred sticking to the postgresql ordering as much as possible.

roji replied:
OK, makes sense.

rwasef1830 commented on src/EFCore.PG/FullTextSearch/NpgsqlFullTextSearchDbFunctionsExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+        /// configuration specified by <paramref name="config" />.
+        /// http://www.postgresql.org/docs/current/static/textsearch-controls.html#TEXTSEARCH-PARSING-QUERIES
+        /// </summary>
+        public static NpgsqlTsQuery PlainToTsQuery(this DbFunctions _, string config, string query) =>
```
See my reply above.

roji replied:
OK

rwasef1830 commented on src/EFCore.PG/FullTextSearch/NpgsqlFullTextSearchDbFunctionsExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+        /// and using the text search configuration specified by <paramref name="config" />.
+        /// http://www.postgresql.org/docs/current/static/textsearch-controls.html#TEXTSEARCH-PARSING-QUERIES
+        /// </summary>
+        public static NpgsqlTsQuery PhraseToTsQuery(this DbFunctions _, string config, string query) =>
```
See my reply above.

roji replied:
OK

rwasef1830 commented:
@roji  see my replies.

rwasef1830 commented:
@roji Rebased and pushed with fixes as reviewed.

Note: I wasn't able to get the BinaryExpression / UnaryExpression idea to work because the built-in classes are sealed tight with internal constructors and argument validation that prevents using these expressions on arbitrary types (must have declared explicit, or implicit conversion operators).

roji commented on src/EFCore.PG/FullTextSearch/NpgsqlFullTextSearchLinqExtensions.cs:
```diff
@@ -0,0 +1,269 @@
+        /// return a new weighted tsvector.
+        /// http://www.postgresql.org/docs/current/static/textsearch-features.html#TEXTSEARCH-MANIPULATE-TSVECTOR
+        /// </summary>
+        public static NpgsqlTsVector SetWeight(
```
OK

roji commented on src/EFCore.PG/FullTextSearch/NpgsqlFullTextSearchLinqExtensions.cs:
```diff
@@ -0,0 +1,269 @@
+        /// return a new weighted tsvector.
+        /// http://www.postgresql.org/docs/current/static/textsearch-features.html#TEXTSEARCH-MANIPULATE-TSVECTOR
+        /// </summary>
+        public static NpgsqlTsVector SetWeight(this NpgsqlTsVector vector, char weight, string[] lexemes) =>
```
OK

roji commented on src/EFCore.PG/FullTextSearch/NpgsqlFullTextSearchDbFunctionsExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+        /// <see cref="PlainToTsQuery(DbFunctions, string, string)" /> instead.
+        /// http://www.postgresql.org/docs/current/static/textsearch-controls.html#TEXTSEARCH-PARSING-QUERIES
+        /// </summary>
+        public static NpgsqlTsQuery ToTsQuery(this DbFunctions _, string config, string query) =>
```
OK

roji commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -141,6 +141,7 @@ protected override Expression VisitUnary(UnaryExpression expression)
                 return expression;
             }
 
+
```
Please remove this extra line

roji commented:
To make it easy to locate, apart from the two whitespace/line nitpicks there's still https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/315/files#r173125096 and https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/315#discussion_r173612303.

roji commented:
One last comment - if you're up to it you can also do some documentation on this. I think at the very least we should update http://www.npgsql.org/efcore/mapping-and-translation.html#operation-translation-to-sql (not necessary a line for each and every operation), or even a separate full-text section that points to the PostgreSQL docs, gives a basic example, etc. A note in [the release notes](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/dev/doc/migration/2.1.md) would also be great.

If you don't have time (or don't like writing docs) I can do it...

rwasef1830 commented:
@roji Removed the cast method and all its relevant code, fixed the 2 nits and force pushed. I rebased on latest dev but I am worried about having messed up BuiltInDataTypesNpgsqlTest because there was a pretty huge merge conflict there. I would appreciate double checking that those tests are correct (they pass except for one that fails due to different ansi encoding on my machine it seems).

About documentation, sure I will update the release notes with a line. Don't know what to edit to make the changes though...

roji commented:
Thanks @rwasef1830, will take a look ASAP. Yeah, I redid a lot of BuiltInDataTypesNpgsqlTest to sync it with EF Core, will take a look at your changes and make sure everything is OK.

Regarding the docs, it's a bit confusing but the EF Core docs are simply in [the doc directory of this repo](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/tree/dev/doc). The website is generated out of [the doc repo](https://github.com/npgsql/doc) which includes the doc directories from npgsql, EF Core and EF6 - this allows us to have a single site consolidating docs from all three projects. Long story short, just make your changes in the doc directory, I'll take of the rest after it's merged.

rwasef1830 commented:
@roji Added some documentation.

rwasef1830 commented on src/EFCore.PG/Query/Internal/NpgsqlEvaluatableExpressionFilter.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Microsoft.EntityFrameworkCore.Query.Internal
+{
+    public class NpgsqlEvaluatableExpressionFilter : RelationalEvaluatableExpressionFilter
```
> Is this because trying to call the full-text search methods on a local vector+query will throw a NotSupportedException

Yes.

It's because of the query compiler. It tries to eliminate all calls in LINQ expressions and convert them to constant expressions (evaluated client side) when they don't reference any column, but in case of full text search functions, there is simply no implementation to run client side, so the behavior is incorrect as the expressions would throw and I must force it to evaluate them server-side. 

```RelationalEvaluatableExpressionFilter``` comes from ```EvaluatableExpressionFilter``` which is a public API from Relinq (not EF Core), so this API will never go away, if EF ever make ```RelationalEvaluatableExpressionFilter``` internal or private, we can easily update and wrap/proxy the original object without any redesign (unless they rebase the entire EF core framework on something other than Relinq but that is highly unlikely). So I think my use of it is the 100% intended use case by Relinq folks and it should be safe to keep.

Sql Server folks have made a choice to let exceptions be thrown if the expression doesn't reference any column (like Freetext) because they guess nobody will use it that way (as 'compute'), but in my opinion that is a bug, or perhaps there is no way in Sql Server to call such functions without referencing columns. Their handling of DateTime.Now is the bare minimum implementation and I think more methods should be added to the filter (anything that doesn't have a client-side implementation).

Allowing users to use the relational store for 'compute' may not be an orthodox use case but there might be some fringe uses for it. I don't see why we should arbitrarily create an incorrect behavior/implementation hole that was not there in EF6 (which can handle such queries just fine) just because the Sql server people decided to do that (or maybe didn't test that case).

Let me know what you think :-).

rwasef1830 replied:
See https://github.com/aspnet/EntityFrameworkCore/issues/2667 @roji when they started using this API from relinq.

roji replied:
@ajcvickers, @divega could we get a bit of guidance here?

To summarize the issue, @rwasef1830 is finalizing the full-text support, and proposes to add NpgsqlEvaluatableExpressionFilter in order to ensure that the relevant functions always execute on the server. This is because these functions, like SQL Server's `FREETEXT`, have no client implementation and would always throw. The main (only?) use case would be to allow running these database functions on client-local data.

My personal opinion is that the usage scenario isn't important enough, but please see @rwasef1830's comment above (and mine). It would be interesting to hear your thoughts on this.

divega replied:
@roji we discussed this with @ajcvickers and @smitpatel (@anpete was not part of the discussion and he may also have something relevant to add). 

In summary, we believe that the behavior @rwasef1830 is proposing is reasonable. We currently are not doing this with the SQL Server provider but in the future we could stop throwing the exception and start issuing new queries.

It will result in N+1 queries being sent to the server, but as long as there are appropriate client evaluation warnings are being reported this is not different from other cases in which other constructs (e.g. subqueries) result into N+1.

There might be a problem with async queries: ideally each individual function evaluation should be evaluated asynchronously not block. I actually haven't looked at the proposed implementation so I am not sure if this is already handled. If it isn't, this is something that I believe @anpete knows how to handle the async composition and he may be able to provide some pointers.

I have created https://github.com/aspnet/EntityFrameworkCore/issues/11466 to track the idea of doing this more often.

roji replied:
OK, thanks for the guidance. I'll merge this PR as-is, and we'll change the implementation when #11466 is implemented (possibly with a better way to add methods as @anpete suggests).

rwasef1830 commented:
@roji Force pushed, added a couple of tests in ```FullTextSearchDbFunctionsNpgsqlTest``` and updated namespace names and commented on the evaluatable expression filter (probably got hidden in the "outdated" conversations in github due to the force push).

roji commented:
OK, great @rwasef1830 and thanks for all your patience. Let's wait and see what the EF Core folks say regarding the evaluatable expression filter question. Other than that I'll merge and amend myself if I run into any more nits.

smitpatel commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlFullTextSearchMethodTranslator.cs:
```diff
@@ -0,0 +1,266 @@
+using Microsoft.EntityFrameworkCore.Query.Expressions.Internal;
+using NpgsqlTypes;
+
+namespace Microsoft.EntityFrameworkCore.Query.ExpressionTranslators.Internal
```
Namespace! (may be other files too)

roji replied:
Yep, this PR was written before The Great Namespace Change, will fix, thanks.

roji commented:
> Btw, the project is depending on nuget package versions that are only available on aspnetcore myget repo (not nuget).

If you're talking about EF Core packages then that's normal - we're tracking 2.1.0-preview2 which is still on myget only. Unfortunately the upcoming enum support which I'm about to merge as well (see branch `enum`) currently depends on Npgsql 4 so that's going to probably be another reference to myget.

roji commented:
Merged to dev via 05cedfdde0e70e0ee80b21488a24eead7d467e25 (and f35bed978f3fc8071c3e31d04dd020916fb6786e), thanks for all this work @rwasef1830! It's great to have PostgreSQL full-text search support!

One question: how would you feel about an overload of `Matches()` that accepts a string, an implicitly does `to_tsquery` to convert it to a query? I'm asking because looking at [this test code](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/05cedfdde0e70e0ee80b21488a24eead7d467e25/test/EFCore.PG.FunctionalTests/Query/FullTextSearchDbFunctionsNpgsqlTest.cs#L696, things seem needlessly verbose with EF.Functions everywhere. There could be other functions where such overloads could make sense.

Also, can you please open an issue to track support for `array_to_tsvector`, just so we have it in the backlog somewhere?

roji commented:
One last thought is that it may be worthwhile to write a separate page describing how to do full-text search, with complete examples (including setting up the index). As it is there's only mention that certain functions are map, but users will likely be a bit bewildered...

Any chance you'd be up for writing that up? If not I'll look into it before release.

rwasef1830 commented:
@roji 
> One question: how would you feel about an overload of Matches() that accepts a string, an implicitly does to_tsquery to convert it to a query? I'm asking because looking at [this test code] things seem needlessly verbose with EF.Functions everywhere. There could be other functions where such overloads could make sense.

I considered this, but decided not to do it because there are native overloads for ```@@``` that accept strings but the semantics are somewhat surprising:

```
The @@ operator also supports text input, allowing explicit conversion of a text string to tsvector or tsquery to be skipped in simple cases. The variants available are:

tsvector @@ tsquery
tsquery  @@ tsvector
text @@ tsquery
text @@ text

The first two of these we saw already. The form text @@ tsquery is equivalent to to_tsvector(x) @@ y. The form text @@ text is equivalent to to_tsvector(x) @@ plainto_tsquery(y).
```

I didn't implement the overload that takes text in place of tsvector because the ```Matches``` C# method stub is implemented as an extension on ```NpgsqlTsVector``` (I would have to make an extension method for string which I didn't want to do because of extension method pollution). 

I did implement an overload which takes string for ```Matches``` for the ```tsquery``` parameter but I send it as-is to postgresql so that it can convert it using ```plain_totsquery```. Doing an implicit ```to_tsquery``` would introduce a semantic difference between us and the native behavior.

rwasef1830 commented:
@roji 
> One last thought is that it may be worthwhile to write a separate page describing how to do full-text search, with complete examples (including setting up the index). As it is there's only mention that certain functions are map, but users will likely be a bit bewildered...

I actually have a couple of extension methods for migrations (for creating the update trigger and the index), but not sure where I can put them. Do we have a samples project ?

roji commented:
@rwasef1830 take a look at the builder extension methods (e.g. `NpgsqlEntityTypeBuilderExtensions`, `NpgsqlIndexBuilderExtensions`). An example is `ForNpgsqlUseXminAsConcurrencyToken`.

Some builder extension methods to simplify the setup sound like a great idea.
